tests/test_checkpointing.py

```diff
@@ -0,0 +1,126 @@
+import logging
+
+import pytest
+import retrying
+import sdk_auth
+import sdk_cmd
+import sdk_tasks
+import shakedown
+import spark_utils as utils
+
+from tests.fixtures.hdfs import SPARK_SUBMIT_HDFS_KERBEROS_ARGS, HDFS_CLIENT_ID, GENERIC_HDFS_USER_PRINCIPAL
+from tests.fixtures.hdfs import hdfs_cmd
+from tests.fixtures.kafka import KAFKA_SERVICE_NAME, KAFKA_PACKAGE_NAME
+from tests.fixtures.kafka import get_kerberized_kafka_spark_conf, upload_jaas
+
+LOGGER = logging.getLogger(__name__)
+HDFS_CHECKPOINT_DIR = "hdfs://hdfs/checkpoints"
+HDFS_TMP_DIR = "/tmp"
+KAFKA_TEST_TOPIC = "streaming_test"
+SPARK_APPLICATION_NAME = "StructuredStreamingWithCheckpointing"
+SPARK_SECURITY_PROTOCOL = "SASL_PLAINTEXT"
+
+def setup_hdfs_paths():
+    sdk_auth.kinit(HDFS_CLIENT_ID, keytab="hdfs.keytab", principal=GENERIC_HDFS_USER_PRINCIPAL)
+    hdfs_cmd("rm -r -skipTrash {}".format(HDFS_CHECKPOINT_DIR))
+    hdfs_cmd("mkdir {}".format(HDFS_CHECKPOINT_DIR))
+    hdfs_cmd("chmod 1777 {}".format(HDFS_CHECKPOINT_DIR))
+
+    hdfs_cmd("rm -r -skipTrash {}".format(HDFS_TMP_DIR))
+    hdfs_cmd("mkdir {}".format(HDFS_TMP_DIR))
+    hdfs_cmd("chmod 1777 {}".format(HDFS_TMP_DIR))
+
+
+def feed_sample_data(jar_uri, kafka_brokers, topic, common_args, messages):
+    producer_args = ["--class", "KerberizedKafkaProducer"] + common_args
+    producer_id = utils.submit_job(app_url=jar_uri,
+                                   app_args="{} {} {} {}".format("kafka", kafka_brokers, topic, ' '.join(messages)),
+                                   service_name=utils.SPARK_SERVICE_NAME,
+                                   args=producer_args)
+
+    # validating producer output
+    utils.check_job_output(producer_id, "{} messages sent to Kafka".format(len(messages)))
+
+
+@pytest.mark.skipif(not utils.hdfs_enabled(), reason='HDFS_ENABLED is false')
+@pytest.mark.sanity
+def test_structured_streaming_recovery(kerberized_spark, kerberized_kafka):
+
+    @retrying.retry(
+        wait_fixed=5000,
+        stop_max_delay=600 * 1000,
+        retry_on_result=lambda res: not res)
+    def wait_for_running_job_output(task_id, expected_line):
+        stdout = sdk_cmd.run_cli("task log --lines=1000 {}".format(task_id))
+        result = expected_line in stdout
+        LOGGER.info('Checking for {} in STDOUT:\n{}\nResult: {}'.format(expected_line, stdout, result))
+        return result
```

Comment:
This looks mostly like [spark_utils.check_job_output](https://github.com/mesosphere/spark-build/blob/master/spark-testing/spark_utils.py#L164) with some retries. Can you move this method to that class or extend the code?

Reply:
the implementation is a bit different for checking the output of running vs completed tasks so I moved this method to `spark_utils`

Reply:
okay